test/python/test_localjob.py

```diff
@@ -15,188 +15,146 @@
-            time.sleep(1)
-
-        # Wait for all the jobs to finish.
-        _ = [job.result() for job in job_array if not job.cancelled]
+
+        executor.submit.assert_called_once()
+        mockedFuture = executor.submit.return_value
+        mockedFuture.cancel.assert_called_once()
 
     def test_done(self):
-        backend = self._provider.get_backend('local_qasm_simulator_py')
-        qobj = qiskit._compiler.compile(self._qc, backend)
-        quantum_job = QuantumJob(qobj, backend, preformatted=True)
-        job = backend.run(quantum_job)
-        job.result()
-        self.assertTrue(job.done)
-
-    def test_get_backend_name(self):
-        backend_name = 'local_qasm_simulator_py'
-        backend = self._provider.get_backend(backend_name)
-        qobj = qiskit._compiler.compile(self._qc, backend)
-        quantum_job = QuantumJob(qobj, backend, preformatted=True)
-        job = backend.run(quantum_job)
-        self.assertTrue(job.backend_name == backend_name)
+        # Like before.
+        # pylint: disable=redefined-outer-name
+        with intercepted_executor_for_LocalJob() as (LocalJob, executor):
+            job = LocalJob(lambda: None, FakeQJob())
+            _ = job.done
+
+        executor.submit.assert_called_once()
+        mockedFuture = executor.submit.return_value
+        mockedFuture.done.assert_called_once()
+
+
+class FakeQJob():
+    def __init__(self):
+        self.backend = FakeBackend()
+        self.qobj = {
+            'id': 'test-id',
+            'config': {
+                'backend_name': self.backend.name,
+                'shots': 1024,
+                'max_credits': 100
+            },
+            'circuits': [{
+                'compiled_circuit_qasm': 'fake-code',
+                'config': {
+                    'seed': 123456
+                }
+            }]
+        }
+
+
+class FakeBackend():
+
+    def __init__(self):
+        self.name = 'test-backend'
+
+
+@contextmanager
+def intercepted_executor_for_LocalJob():
+    """Context that patches the derived executor classes to return the same
+    executor object. Also patches the future object returned by executor's
+    submit()."""
+
+    import importlib
+    import concurrent.futures as futures
+    import qiskit.backends.local.localjob as localjob
+
+    executor = unittest.mock.MagicMock(spec=futures.Executor)
+    executor.submit.return_value = unittest.mock.MagicMock(spec=futures.Future)
+    mock_options = {'return_value': executor, 'autospec': True}
+    with patch.object(futures, 'ProcessPoolExecutor', **mock_options),\
+            patch.object(futures, 'ThreadPoolExecutor', **mock_options):
+        importlib.reload(localjob)
+        yield localjob.LocalJob, executor
+
+
+@contextmanager
+def mocked_simulator_binaries():
+    """Context to force binary-based simulators to think the simulators exist.
+    """
+    from qiskit.backends.local import qasm_simulator_projectq
```

Comment:
Was the choice of the binary simulator here arbitrary? If so, it may be better to use a non-third-party one.

Reply:
Right although, technically, I'm not testing the simulators but [all the backends that come with QISKit](https://github.com/QISKit/qiskit-sdk-py/tree/master/qiskit/backends/local) regardless of which simulator they are interfacing with. That's why I'm patching the objects pretending the simulators are there. Does it make sense?

Reply:
I think I see what this is doing now.

Reply:
I think the implementation is neat, although the intent might fall a bit outside the principles applied to the rest of the suite - in general we rely on the fact that if a backend cannot be instantiated, it means is "optional" or not available on the current environment: however, during the second stage of Travis we do ensure the cpp sims are available anyway.

So in practice the mocking would only affect the `projectq` simulator most of the time (which is bound to be removed from the core as well), and might be a bit tricky to keep the mocking robust enough depending on how the codebase evolves - it is fine for me to leave it as it is, but might make sense to just simplify and avoid the mocking  in the future :+1:

Reply:
So, do you prefer the tests for these cases to be optional?